src/pybind/mgr/nfs/export.py

```diff
@@ -496,46 +518,44 @@ def get_export_by_pseudo(
         export = self._fetch_export(cluster_id, pseudo_path)
         return export.to_dict() if export else None
 
-    def apply_export(self, cluster_id: str, export_config: str) -> Tuple[int, str, str]:
+    def apply_export(self, cluster_id: str, export_config: str) -> AppliedExportResults:
```

Comment:
It looks nicer but I think that the expectation is that NotImplementedError is raised by _apply_export. Specificcally, I think that the rook module is enabled _update_export may raise NotImplementedError because it's (potentially) hooking back into the orchestration layer.... when need_nfs_service_restart is true.

It would have been much clearer if a more specific exception type had been defined though. Because most of this is a guess.  If we want to retain the actual nesting but make it easier to read we could split apply_export into two functions:

```python
# pseudocode
def apply_export(...):
    try:
        return self._apply_all_exports(cluster_id, export_config)
    except NotImplementedError:
        raise ...
    except blah:
        raise ....
        
        
def _apply_all_exports(...):
    aeresults = AppliedExportResults()
    for export in ...:
      ... blah blah ...
    return aeresults
```

Does that solve both issues?

Reply:
> It looks nicer but I think that the expectation is that NotImplementedError is raised by _apply_export. Specificcally, I think that the rook module is enabled _update_export may raise NotImplementedError because it's (potentially) hooking back into the orchestration layer.... when need_nfs_service_restart is true.
> 

Thanks! I missed thinking about the rook implementation. Can we add a comment above "except NotImplementedError:" explaining why we do this?

> It would have been much clearer if a more specific exception type had been defined though. Because most of this is a guess. If we want to retain the actual nesting but make it easier to read we could split apply_export into two functions:
> 
> ```python
> # pseudocode
> def apply_export(...):
>     try:
>         return self._apply_all_exports(cluster_id, export_config)
>     except NotImplementedError:
>         raise ...
>     except blah:
>         raise ....
>         
>         
> def _apply_all_exports(...):
>     aeresults = AppliedExportResults()
>     for export in ...:
>       ... blah blah ...
>     return aeresults
> ```
> 
> Does that solve both issues?

Even here I assume you'd be catching all exception raised by _apply_export() and neatly putting it into aerresults. How is the NotImplementedError propagated to the outer try catch block?

Reply:
I added a comment.

Unfortunately, I'm not clear on what you are asking in the 2nd part. The call chain is roughly: apply_export -> _apply_export -> _update_export today. If we want to catch the NotImplemtedError we have to have the except clause around all the calls to _apply_export.

Since you already approved below I think, if it's OK with you, that I'll leave this in place for now and if you would like to talk about this some more we can discuss it over a tracker issue or in the orchestration weekly meeting(s) and I'll follow up in a later PR. Sound all right?

Reply:
Discussed in https://github.com/ceph/ceph/pull/46209#discussion_r946253434